aptos-move/framework/aptos-framework/sources/vesting_without_staking.move

```diff
@@ -0,0 +1,1132 @@
+    public entry fun vest(contract_address: address) acquires VestingContract {
+        assert_active_vesting_contract(contract_address);
+        let vesting_contract = borrow_global_mut<VestingContract>(contract_address);
+        // Short-circuit if vesting hasn't started yet.
+        if (vesting_contract.vesting_schedule.start_timestamp_secs > timestamp::now_seconds()) {
+            return
+        };
+
+        // Check if the next vested period has already passed. If not, short-circuit since there's nothing to vest.
+        let vesting_schedule = &mut vesting_contract.vesting_schedule;
+        let last_vested_period = vesting_schedule.last_vested_period;
+        let next_period_to_vest = last_vested_period + 1;
+        let last_completed_period =
+            (timestamp::now_seconds() - vesting_schedule.start_timestamp_secs) / vesting_schedule.period_duration;
+        if (last_completed_period < next_period_to_vest) {
+            return
+        };
+
+        // Index is 0-based while period is 1-based so we need to subtract 1.
+        let schedule = &vesting_schedule.schedule;
+        let schedule_index = next_period_to_vest - 1;
+        let vesting_fraction = if (schedule_index < vector::length(schedule)) {
+            *vector::borrow(schedule, schedule_index)
+        } else {
+            // Last vesting schedule fraction will repeat until the grant runs out.
+            *vector::borrow(schedule, vector::length(schedule) - 1)
+        };
+
+        vesting_schedule.last_vested_period = next_period_to_vest;
+
+        emit_event(
+            &mut vesting_contract.vest_events,
+            VestEvent {
+                admin: vesting_contract.admin,
+                vesting_contract_address: contract_address,
+                period_vested: next_period_to_vest,
+            },
+        );
+        // Every shareholder should receive the money in proportion to their shares.
+        distribute(vesting_fraction, contract_address);
+    }
+
+    /// Distribute any withdrawable grant.
+    /// This is no entry function anymore as it's called from within the vest.
+    fun distribute(vesting_fraction: FixedPoint32, contract_address: address) acquires VestingContract {
+        assert_active_vesting_contract(contract_address);
+
+        let vesting_contract = borrow_global_mut<VestingContract>(contract_address);
+        let vesting_signer = get_vesting_account_signer_internal(vesting_contract);
+        let shareholders_address = simple_map::keys(&vesting_contract.shareholders);
+        let total_amount_left = 0;
+        // Distribute coins to shareholders.
+        vector::for_each_ref(&shareholders_address, |shareholder| {
+            let shareholder = *shareholder;
+            let amount = min(simple_map::borrow(& vesting_contract.shareholders, &shareholder).left_amount,fixed_point32::multiply_u64(simple_map::borrow(& vesting_contract.shareholders, &shareholder).init_amount, vesting_fraction));
```

Comment:
Here also `borrow` is O(n) so the entire loop require O(n^2), instead you may use `simple_map::to_vec_pair` so that you need to iterate the vectors only once, Since the returned vectors can be destroy, you can use vector::pop while iterating which is cheaper in gas cost as compared to vector::borrow

Reply:
Note that `borrow_mut` is still required below at line `414` since otherwise we can't make in-place modification.

Reply:
Updated!